crates/build/src/docker.rs

```diff
@@ -0,0 +1,302 @@
+            .manifest_path
+            .absolute_directory()?
+            .as_path()
+            .to_owned();
+
+        let file_path = host_folder.join(Path::new("target/build_result.json"));
+
+        let mount = Mount {
+            target: Some(String::from("/contract")),
+            source: Some(
+                host_folder
+                    .to_str()
+                    .context("Cannot convert path to string.")?
+                    .to_string(),
+            ),
+            typ: Some(MountTypeEnum::BIND),
+            ..Default::default()
+        };
+        println!("Mount cfg: {:?}", &mount);
+        let host_cfg = Some(HostConfig {
+            mounts: Some(vec![mount]),
+            ..Default::default()
+        });
+
+        let entrypoint = vec!["/bin/bash", "-c"]
+            .iter()
+            .map(|s| s.to_string())
+            .collect();
+
+        let mut args: Vec<String> = Vec::new();
+
+        features.append_to_args(&mut args);
+        if keep_debug_symbols {
+            args.push("--keep-debug-symbols".to_owned());
+        }
+
+        if let Some(passes) = optimization_passes {
+            if passes != OptimizationPasses::default() {
+                args.push(format!("--optimization-passes {}", passes));
+            }
+        }
+
+        if network == Network::Offline {
+            args.push("--offline".to_owned());
+        }
+
+        if build_mode == BuildMode::Release {
+            args.push("--release".to_owned());
+        }
+
+        if unstable_flags.original_manifest {
+            args.push("-Z original-manifest".to_owned());
+        }
+
+        let s = match build_artifact {
+            BuildArtifacts::CodeOnly => "--generate code-only".to_owned(),
+            BuildArtifacts::All => String::new(),
+            BuildArtifacts::CheckOnly => {
+                anyhow::bail!("--generate check-only is invalid flag for this command!");
+            }
+        };
+
+        args.push(s);
+
+        if max_memory_pages != DEFAULT_MAX_MEMORY_PAGES {
+            args.push(format!("--max-memory-pages {}", max_memory_pages));
+        }
+
+        let joined_args = args.join(" ");
+
+        let cmds = vec![format!(
+            "cargo contract build {} --output-json > target/build_result.json",
+            joined_args
+        )];
+
+        let config = Config {
+            image: Some(build_image.id.clone()),
+            entrypoint: Some(entrypoint),
+            cmd: Some(cmds),
+            host_config: host_cfg,
+            attach_stderr: Some(true),
+            ..Default::default()
+        };
+
+        let container_id = match client
+            .create_container(options.clone(), config.clone())
+            .await
+        {
+            Ok(response) => response.id,
+            Err(_) => {
+                // container might exist, so we delete the previous one
+                let _ = client.remove_container("ink-container", None).await;
```

Comment:
Is this going download the dependencies every time it's re-run? even for the same contract?

Since we're not mounting the local `~/.cargo` folder it will have to download and compile dependencies inside docker. If you kill the previously-run container it will have to do it again each time.

Reply:
If the container compiled the contract before, and there were no changes in the metadata or code, then it will not rebuild the contract. Otherwise yes, everytime the command is executed, a new container is built and all ink! dependencies are pulled from scratch. I was thinking of reusing the same container for the contract it was made, but then it will bloat the local storage. 

I might enable container reuse by default and then add the flag to remove the container after the build

Reply:
Another option is to start the container in the background (`docker run`) and then send build commands to it (`docker exec`). If no container is running - start new one as daemon, if it's already running: reuse. You can set a TTL on it and the system would kill it after specified time.

Reply:
You still need to create a new container for each contract as you cannot change the mounting config. Even having the TTL, I am not sure what are the implications of having a running idle container on the performance of external system that's gonna run the command.

Reply:
exit code 137 means that your system killed the container b/c it tried to use more memory that it was allowed to.